### DISCUSSION TITLE
Can't Compile The Latest Version of the cardano-wallet with cabal errors out: cabal: Error parsing project file (line 20)

### DISCUSSION THREAD

**Discussion by jthompsonic** [cardano-foundation/cardano-wallet#4616] [General]
I'm a newbie compiling the cardano-wallet from source in Ubuntu 24.04.  I already have compiled a cardano node from source and it's working.  

I cloned the latest version of cardano-wallet: git show
commit 7fc14309fda297f74eec491b4bb45bc9eb42158e (HEAD -> master, origin/master, origin/HEAD)
Merge: 9fd1d0ffe3 8cee4c63ed
Author: Paolo Veronelli <paolo.veronelli@gmail.com>
Date:   Tue Jun 4 15:43:11 2024 +0200

    [ADP-3366] Refactor genesis files creation in local cluster (#4611)
    
    - [x] refactor GenesisFiles creation
    - [x] move the GenesisFiles module inside the Node namespace
    
    ADP-3366
======================

I'm following the compile instructions at https://cardano-foundation.github.io/cardano-wallet/contributor/what/building.html and here is the error I run into (You can see I included the versions of my cabal / ghc tools):

~/cardano-wallet$ cabal update
cabal: Error parsing project file
cardano-wallet/cabal.project:178:
"<eitherParsec>" (line 20, column 1):
unexpected ","
expecting space, "+" or "-"

~/cardano-wallet$ cabal --version
cabal-install version 3.6.2.0
compiled using version 3.6.2.0 of the Cabal library 
~/cardano-wallet$ ghc --version
The Glorious Glasgow Haskell Compilation System, version 9.2.8
~/cardano-wallet$ nix --version
nix (Nix) 2.18.1

Any help on getting this to compile would be greatly appreciated.  Thanks.

**Comment by paolino** [comment]
It looks like that specific commit at least got a green from CI after merging to master. So it should build.

Can you try `nix build` ?

**Comment by paolino** [comment]
Another thing is we do not support 9.2.8 as a compiler due to performance regressions. With `nix build`  you will compile with 9.6.4.
Is there a specific reason why you cannot use the released artifacts in https://github.com/cardano-foundation/cardano-wallet/releases/tag/v2024-05-05 ?

**Comment by jthompsonic** [comment]
I switched ghc to 9.6.4 and I get the same parser error on the contstraints line in cabal.project (line 178). I have looked at the cabal.project and associated documentation and I don't see an obvious configuration error. So still not getting very far in the complilation process for cardano-wallet.

I tried the nix build system.  It's too big of an endeavor for me to use that system. I much prefer smaller and understandable (build) systems.

I did try the source code at https://github.com/cardano-foundation/cardano-wallet/releases/tag/v2024-05-05 and it has the same problem with the same cabal parse build error.  I thought maybe it got fixed in your latest git version but that is not the case.  

When you use the word "artifacts" which can be applied to the compiled version instead of the source code, for security review reasons, I prefer to build from source, so I can inspect how work is getting done to satisfy my own curiosity and learning process. I like to open the trunk and look under the hood in addition to kicking the tires. 

I would appreciate if you can let me know when a successful build works according to your build instructions.

**Reply by paolino** [reply]
Those building instructions are not up-to-date , sadly. I asked around and none of us maintainers is building outside a nix shell.
I personally have a special distaste for nix, mostly because I don't understand it probably. 
I'm also surprised by the error itself, so I will take a closer look at that anyway.
We created 2 tickets which , sadly again, you cannot see outside of CF.
- one to remove the instruction for building outside nix
- one to document how to build a statically linked version of the executable of the actual code

The first ticket will fail if we find a way to refresh the instructions. I will have a go at that and ask around.
The second ticket should be easy because we already do that when we create a release artifact. Still it will be a nix expression, so you will need a nix environment to run it.

I personally have a special distaste for nix, mostly because I don't understand it probably. 
But there are at least some good things about nix governing the builds. 

I'm not an expert whatsoever, so there could be much more.

With nix you will have somewhat the exact same tools as we use to work on this code (ghc, cabal, hls, stylish-haskell, ...)
This is cool because tickets like this are not necessary anymore (if you have a correct set up of nix :-] ) and there is a higher chance to create patches that will pass the CI and the reviews, yay!

With nix you will have the same non-haskell libraries to link against (libsodium IIRC is somewhat specific, but there could be more, not sure). This is great that you don't have to hunt them in the host packages, or eventually build them yourself which recursively send us back to square 0 ...

Given that, nix is an elephant jumping between the storage and the network. 
1. Better keep it in a separate partition and be sure you have the cache set up correctly or you will end up recompiling ghc. (Or if you want to compile ghc , you can exactly do that (a bit boring achievement)) . 
2. Better avoid pay per use connections or volume constrained ones.
3. Better run nix-collect-garbage, quite often. 
4. Better have something to do while it's his turn to refresh the environment, especially because we still don't have a cache of our nix evaluations, it will recompile a decent part of the haskell ecosystem on unlucky days.

Btw, here is my nix configuration, in case you want to embark on this.

```
build-users-group = nixbld
substituters         = https://cache.iog.io https://cache.nixos.org
trusted-substituters =
trusted-public-keys  = hydra.iohk.io:f/Ea+s+dFdN+3Y/G+FDgSq+a5NEWhJGzdjvKNGv0/EQ= cache.nixos.org-1:6NCHdD59X431o0gWypbMrAURkbJ16ZPMQFGspcDShjY=
experimental-features = nix-command flakes
max-jobs = 30
trusted-users = paolino
```
HTH

I will update here, once / if I got the compilation outside the nix in a better shape

**Reply by paolino** [reply]
I can get past your point, out of the nix shell. It seems you need a newer cabal-install. It works at least with 3.10.3. And it doesn't at least with 3.6.2.0
```
paolino@plutimus:~/iohk/cardano-wallet$ ghcup  set ghc 9.6.4                                                         
[ Info  ] downloading: https://raw.githubusercontent.com/haskell/ghcup-metadata/master/ghcup-0.0.8.yaml as file /home/paolino/.ghcup/cache/ghcup-0.0.8.yaml
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
[ Warn  ] New ghc version available. If you want to install this latest version, run 'ghcup install ghc 9.10.1'
[ Warn  ] New stack version available. If you want to install this latest version, run 'ghcup install stack 2.15.7'
[ Warn  ] [GHCup-00330] ghc is shadowed by /home/paolino/.local/bin/ghc.
[ ...   ] The upgrade will not be in effect, unless you remove /home/paolino/.local/bin/ghc
[ ...   ] or make sure /home/paolino/.ghcup/bin comes before /home/paolino/.local/bin in PATH.
[ Info  ] GHC 9.6.4 successfully set as default version
paolino@plutimus:~/iohk/cardano-wallet$ ghcup  set cabal   
[ Warn  ] New ghc version available. If you want to install this latest version, run 'ghcup install ghc 9.10.1'
[ Warn  ] New stack version available. If you want to install this latest version, run 'ghcup install stack 2.15.7'
[ Info  ] Cabal 3.10.3.0 successfully set as default version
paolino@plutimus:~/iohk/cardano-wallet$ cabal freeze
Warning: Requested index-state 2024-06-04T00:00:00Z is newer than
'hackage.haskell.org'! Falling back to older state (2024-06-03T20:42:59Z).
Resolving dependencies...
Error: cabal: Could not resolve dependencies:
[__0] trying: address-derivation-discovery-2024.5.5 (user goal)
[__1] next goal: cardano-crypto-class (dependency of
address-derivation-discovery)
[__1] rejecting: cardano-crypto-class-2.1.4.0 (conflict: pkg-config package
libblst-any, not found in the pkg-config database)
[__1] rejecting: cardano-crypto-class-2.1.3.0, cardano-crypto-class-2.1.2.0,
cardano-crypto-class-2.1.1.0, cardano-crypto-class-2.1.0.2,
cardano-crypto-class-2.1.0.1, cardano-crypto-class-2.1.0.0,
cardano-crypto-class-2.0.0.1, cardano-crypto-class-2.0.0.0.1,
cardano-crypto-class-2.0.0 (constraint from project config 
/home/paolino/iohk/cardano-wallet/cabal.project requires ==2.1.4.0)
[__1] fail (backjumping, conflict set: address-derivation-discovery,
cardano-crypto-class)
After searching the rest of the dependency tree exhaustively, these were the
goals I've had most trouble fulfilling: cardano-crypto-class,
address-derivation-discovery
```

But not much more because of `blst` missing on my debian machine.  I have seen instructions 

- here, https://github.com/IntersectMBO/cardano-node/issues/5378 , for blst

- here, https://github.com/IntersectMBO/cardano-node/issues/5528 , for sodium

I'm not going to try them , at least in the short period. Sadly I will keep on working with the elephant for now. 
HTH

**Reply by jthompsonic** [reply]
First thank you for your response and the time it took for you to do so.  I appreciate that. I understand you have momentum in a project and wanting to keep things going as they have developed into your current project with nix.  As you defined it, nix is a big fat elephant as a build system. It's just too painful for me to use. It breaks my rule of keeping things simple and understandable so I can maintain a level of security and functionality. 

So I choose not to use nix (which may mean I can't do a lot of things in the cardano world ...).  

I'm ok with the complexities of cabal but that's where I draw the line for myself as a developer.  I prefer CMAKE.

I tried to build with cabal v3.10.3.0 and ghc v9.6.4 and got past the initial error I reported but now I have ended up with this error:

|cabal build all --minimize-conflict-set
Warning: Requested index-state 2024-06-04T00:00:00Z is newer than
'hackage.haskell.org'! Falling back to older state (2024-06-03T20:42:59Z).
Resolving dependencies...
Error: cabal: Could not resolve dependencies:
[__0] trying: address-derivation-discovery-2024.5.5 (user goal)
[__1] next goal: ouroboros-consensus-cardano (dependency of
address-derivation-discovery)
[__1] rejecting: ouroboros-consensus-cardano-0.16.0.0,
ouroboros-consensus-cardano-0.15.0.0, ouroboros-consensus-cardano-0.14.2.0
(constraint from project config /home/admin/cardano-wallet/cabal.project
requires ==0.14.1.0)
[__1] trying: ouroboros-consensus-cardano-0.14.1.0
[__2] next goal: ouroboros-consensus-diffusion (dependency of
ouroboros-consensus-cardano)
[__2] rejecting: ouroboros-consensus-diffusion-0.16.0.0,
ouroboros-consensus-diffusion-0.15.0.0 (constraint from project config
/home/admin/cardano-wallet/cabal.project requires ==0.14.0.0)
[__2] trying: ouroboros-consensus-diffusion-0.14.0.0
[__3] trying: cardano-crypto-tests-2.1.2.0 (dependency of
ouroboros-consensus-diffusion)
[__4] next goal: cardano-crypto-praos (dependency of cardano-crypto-tests)
[__4] rejecting: cardano-crypto-praos-2.1.2.0 (conflict: pkg-config package
libsodium-any, not found in the pkg-config database)
[__4] rejecting: cardano-crypto-praos-2.1.1.2, cardano-crypto-praos-2.1.1.1,
cardano-crypto-praos-2.1.1.0, cardano-crypto-praos-2.1.0.0,
cardano-crypto-praos-2.0.0.1, cardano-crypto-praos-2.0.0.0.1,
cardano-crypto-praos-2.0.0 (constraint from project config
/home/admin/cardano-wallet/cabal.project requires ==2.1.2.0)
[__4] fail (backjumping, conflict set: cardano-crypto-praos,
cardano-crypto-tests)
After searching the rest of the dependency tree exhaustively, these were the
goals I've had most trouble fulfilling: ouroboros-consensus-diffusion,
cardano-crypto-praos, cardano-crypto-tests, ouroboros-consensus-cardano,
address-derivation-discovery

Thanks for any more tips you can give me on the cabal build.

**Reply by ltouro** [reply]
the key part of the error is 'libsodium' in 'libsodium-any, not found in the pkg-config database'. I once took this endeavor and I found some hints on how to proceed here: https://github.com/input-output-hk/hydra/wiki/Logbook-2021-H2#dependencies-and-build. The issue paolino linked above may also be useful. From what I remember, I could get past this point but ended hitting another wall.

This is from 2021, so put that into perspective:
```
git clone https://github.com/input-output-hk/libsodium
cd libsodium/
git checkout 66f017f16633f2060db25e17c170c2afa0f2a8a1
./autogen.sh
./configure
make && sudo make install
```
I wish you good luck! Let me know if you succeed
